Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/bugreporting/loghistory/RollingLogHistory.kt

```diff
@@ -0,0 +1,64 @@
+package de.rki.coronawarnapp.bugreporting.loghistory
+
+import android.util.Log
+import de.rki.coronawarnapp.util.coroutine.AppScope
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.math.min
+
+@Singleton
+class RollingLogHistory @Inject constructor(
+    @AppScope private val scope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider
+) : Timber.DebugTree() {
+
+    private val bufferLock = Mutex()
+    private val buffer: ArrayDeque<String> = ArrayDeque(BUFFER_SIZE + 1)
+    private val logQueue = MutableStateFlow("")
+
+    init {
+        logQueue
+            .filter { it.isNotBlank() }
+            .onEach {
+                bufferLock.withLock {
+                    buffer.addFirst(it)
+                    if (buffer.size > BUFFER_SIZE) {
```

Comment:
If you do ">=" here, you won't need the  "BUFFER_SIZE + 1" above  which will resolve in less magic numbers.

Reply:
adjusted.